### PR TITLE
msp: automatically resolve container image digest

### PR DESCRIPTION
**Pull request by michaellzc**
`sg msp generate <svc> <env>` now automatically resolves image tag to `<tag>@<digest>` in the event if `digest` is missing

use cases:

- run `sg msp generate` can automatically update the service to the latest `insiders` tag and trigger an update cloud run service
- ensure we always pin to a specific digest

## Test plan

See https://github.com/sourcegraph/managed-services/pull/13

### CODE REVIEW

**Comment by michaellzc** [comment]
Current dependencies on/for this PR:
* main
  * **PR #57953** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57953" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57953?utm_source=stack-comment).

File: dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go [hunk 1]
```diff
@@ -354,7 +354,7 @@ type cloudRunServiceBuilder struct {
 
 func (c cloudRunServiceBuilder) Build(stack cdktf.TerraformStack, vars Variables) (cloudrunv2service.CloudRunV2Service, error) {
 	// TODO Make this fancier, for now this is just a sketch of maybe CD?
-	serviceImageTag, err := vars.Environment.Deploy.ResolveTag()
+	serviceImageTag, err := vars.Environment.Deploy.ResolveTag(vars.Image)
```
Comment (bobheadxi) [inline dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:357]: at the definition this is `repo`, but here at callsite it is `image` - should we just ask the resolver to accept an image?

File: dev/managedservicesplatform/spec/environment.go [hunk 2]
```diff
@@ -63,14 +66,24 @@ const (
...
 	switch d.Type {
 	case EnvironmentDeployTypeManual:
-		if d.Manual == nil {
-			return "insiders", nil
+		var tag string
+		if d.Manual == nil || d.Manual.Tag == "insiders" {
+			tag = "insiders"
+		} else {
+			tag = d.Manual.Tag
 		}
```
Comment (bobheadxi) [inline dev/managedservicesplatform/spec/environment.go:77]:
```suggestion
		if d.Manual == nil {
			tag = "insiders"
		} else {
			tag = d.Manual.Tag
		}
```

File: dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go (hunk 1 above)
Comment (bobheadxi) [reply dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:357]: though I guess terminology is pretty loose with images, where the image == the image repo

**Review by bobheadxi — APPROVED** [review]
thank you 🎉 😁

File: dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go (hunk 1 above)
Comment (michaellzc) [reply dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:357]: yeah, the image name spec is extremely loose (hence they just call it the "ref" lol), let's keep it the way it is

File: dev/managedservicesplatform/spec/environment.go [hunk 3]
```diff
@@ -63,14 +66,24 @@ const (
 // ResolveTag uses the deploy spec to resolve an appropriate tag for the environment.
 //
 // TODO: Implement ability to resolve latest concrete tag from a source
-func (d EnvironmentDeploySpec) ResolveTag() (string, error) {
+func (d EnvironmentDeploySpec) ResolveTag(repo string) (string, error) {
 	switch d.Type {
 	case EnvironmentDeployTypeManual:
```
Comment (bobheadxi) [inline dev/managedservicesplatform/spec/environment.go:71]: afterthought - should we introduce another type here?

Comment (michaellzc) [reply dev/managedservicesplatform/spec/environment.go:71]: auto? what’s the desired behaviour?

right now, we always fetch the latest digest for a tag, so even if it’s manual, it’s kind of auto

Comment (bobheadxi) [reply dev/managedservicesplatform/spec/environment.go:71]: IMO, `manual` should never automatically update, and a new type `auto` should have the new behaviour maybe

Comment (michaellzc) [reply dev/managedservicesplatform/spec/environment.go:71]: sgtm, will follow up next week

**Pull request merged**